cmd/build/validation.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2022 The Okteto Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/okteto/okteto/pkg/cmd/build"
+	oktetoErrors "github.com/okteto/okteto/pkg/errors"
+	oktetoLog "github.com/okteto/okteto/pkg/log"
+	"github.com/okteto/okteto/pkg/model"
+	"github.com/okteto/okteto/pkg/okteto"
+	"github.com/okteto/okteto/pkg/registry"
+	"github.com/okteto/okteto/pkg/types"
+	"golang.org/x/sync/errgroup"
+)
+
+// GetServicesToBuild returns the services it has to built because they are not already built
+func (bc *Command) GetServicesToBuild(ctx context.Context, manifest *model.Manifest) ([]string, error) {
+	buildManifest := manifest.Build
+
+	// check if images are at registry (global or dev) and set envs or send to build
+	toBuild := make(chan string, len(buildManifest))
+	g, _ := errgroup.WithContext(ctx)
+	for service := range buildManifest {
+		svc := service
+		g.Go(func() error {
+			return bc.checkServicesToBuild(svc, manifest, toBuild)
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return nil, err
+	}
+	close(toBuild)
+
+	if len(toBuild) == 0 {
+		oktetoLog.Information("Images were already built. To rebuild your images run 'okteto build' or 'okteto deploy --build'")
+		return nil, nil
+	}
+
+	svcsToBuildList := []string{}
+	for svc := range toBuild {
+		svcsToBuildList = append(svcsToBuildList, svc)
+	}
+	return svcsToBuildList, nil
+}
+
+func (bc *Command) checkServicesToBuild(service string, manifest *model.Manifest, ch chan string) error {
+	buildInfo := manifest.Build[service]
+	isStack := manifest.Type == model.StackType
+	if isStack && okteto.IsOkteto() && !registry.IsOktetoRegistry(buildInfo.Image) {
+		buildInfo.Image = ""
+	}
+	opts := build.OptsFromBuildInfo(manifest.Name, service, buildInfo, &types.BuildOptions{})
+
+	if build.ShouldOptimizeBuild(opts.Tag) {
```

Comment:
I think there is some logic similar to this in `buildv2` file. Do they have any difference? We are doing this refactor to avoid to have different build logic. I would like to avoid to have different logic to optimize the builds. Can you double check?

Reply:
Yes, this is intended, one is to prepare for building and the other one to check if it's already built